docs/proposals/20220125-ipam-integration.md

```diff
@@ -0,0 +1,340 @@
+
+In order to consume IPAddresses, infrastructure providers need to create an IPAddressClaim for each address they require, and then have to wait until all claims are fulfilled. When creating those claims, a deterministic naming scheme should be used, so it's easy to identify which claims belong to which machine, e.g. `<infra machine name>-<interface name>`.
+
+Since IPAddressClaims have to reference a Pool, the infrastructure provider needs to allow specifying a Pool in its infrastructure templates, and the referenced Pool also has to be created (or exist already) when creating a cluster. The Pool has to be in the same namespace as the cluster, which should be enforced by not providing a namespace field on the reference.
+
+![relationships of ipam custom resources](images/ipam-integration/consumption.png)
+
+When creating an IPAddressClaim, the provider must set an owner reference to the infrastructure machine on creation, so IPAM providers can fetch further information if necessary.
+
+Using CAPV as an example, a template *could* look like this:
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: VSphereMachineTemplate
+metadata:
+ name: example
+ namespace: vsphere-site1
+spec:
+ template:
+   spec:
+     cloneMode: FullClone
+     numCPUs: 8
+     memoryMiB: 8192
+     diskGiB: 45
+     network:
+       devices:
+       - dhcp4: false
+         fromPool: # reference to the pool
+           group: ipam.cluster.x-k8s.io/v1alpha1
+           kind: IPPool
+           name: testpool
+```
+
+A Machine generated from that template could then look like this:
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: VSphereMachine
+metadata:
+name: example-1
+namespace: vsphere-site1
+spec:
+ cloneMode: FullClone
+ numCPUs: 8
+ memoryMiB: 8192
+ diskGiB: 45
+ network:
+   devices:
+   - dhcp4: false
+     fromPool: # reference to the pool
+       group: ipam.cluster.x-k8s.io/v1alpha1
+       kind: IPPool
+       name: testpool
+status:
+ network:
+   devices:
+     - claim:
+         name: example-1-1
+```
+
+### User Stories
+
+#### Story 1
+
+As a cluster operator I want to use static IP addresses with my Machines to avoid problems with DHCP lease duration during rolling cluster upgrades.
+
+#### Story 2
+
+As a cluster operator I want to integrate machines provisioned by CAPI with an external IPAM system and allocate IP addresses from it.
+
+#### Story 3
+
+As an infrastructure provider I want a single interface to integrate with multiple different IPAM providers.
+
+#### Story 4
+
+As an IPAM provider I want a single interface to integrate with all infrastructure providers that can benefit from IPAM integration.
+
+### Implementation Details/Notes/Constraints
+
```

Comment:
can we talk a bit here about contract/good practices on how consumers i.e infraProviders are expected to interact with this API?
e.g:
- An infraProvider which Machine has an IPPool reference must create a IPClaim for it (and add an ownerRef?).
- An infraProvider owning and IPClaim must wait for the resulting IP until it keeps making progress.
- An infraProvider waiting for an IP must signal that in its conditions and this condition must be required to compute for the "ready" condition with reason="waitingForIP".
- An infraProvider which is given an IP must signal that in its conditions and this condition must be required to compute for the "ready".
- An infraProvider referencing a IPPool must set the given IP in its .status.Addresses.

Reply:
I've added two sections for ipam and infra providers with more details.